tensorforce/models/model.py

```diff
@@ -830,6 +859,26 @@ def tf_preprocess_reward(self, states, internals, terminal, reward):
 
         return reward
 
+    def tf_episode_reward(self, terminal, reward):
+        reach_terminal = terminal[-1]
+        reward_sum = tf.reduce_sum(input_tensor=reward)
+        episode_reward = self.episode_reward.assign_add(delta=reward_sum)
+        return tf.cond(
+            pred=reach_terminal,
+            true_fn=(lambda: self.summary_episode_reward(episode_reward=episode_reward)),
+            false_fn=(lambda: episode_reward)
```

Comment:
I've debug that the error InvalidArgumentError : Retval[0] does not have value is because the false_fn of tf.cond, but I can't figure out what's wrong with it, do you have any idea about it? @michaelschaarschmidt  @AlexKuhnle   Thanks!

Reply:
@yellowbee686 if I recall from the specs of the TF Assign ops, "episode_reward" as you have it above is the Op to do the Assign_add and not the actual resulting value of the addition.  So false_fn is only assigning the value and returning that tf.assign_add() op.  I think you need to use a control flow to do what you are tryinig:

    with tf.control_dependencies([episode_reward]):
        return self.episode_reward

episode_reward is the assign_add Op and not the result.  The result will be in self.episode_reward according to your code.

Reply:
@Mazecreator Thank you very much, I'll try it!

Reply:
def tf_episode_reward(self, terminal, reward):
        reach_terminal = terminal[-1]
        reward_sum = tf.reduce_sum(input_tensor=reward)
        episode_reward = self.episode_reward.assign_add(delta=reward_sum)
        return tf.cond(
            pred=reach_terminal,
            true_fn=(lambda: self.summary_episode_reward(episode_reward=episode_reward)),
            false_fn=(lambda: self.assign_episode_reward(episode_reward=episode_reward))
        )

    def assign_episode_reward(self, episode_reward):
        with tf.control_dependencies(control_inputs=(episode_reward,)):
            return self.episode_reward

    def summary_episode_reward(self, episode_reward):
        if 'episode-reward' in self.summary_labels:
            summary = tf.summary.scalar(name='episode-reward', tensor=episode_reward)
            self.summaries.append(summary)
        return episode_reward

@Mazecreator I've tried as your advice and the error still occurs, here I try to return the assign_op and in observe I'll do the assign op in model.observe by 
episode, _ = self.monitored_session.run(fetches=(self.increment_episode, self.increment_episode_reward), feed_dict=feed_dict)

It's quite weird that if I change the code to
return tf.cond(
            pred=reach_terminal,
            true_fn=(lambda: self.summary_episode_reward(episode_reward=episode_reward)),
            false_fn=(lambda: self.summary_episode_reward(episode_reward=episode_reward))
        )
the error is the same

but if I change the code to:
return tf.cond(
            pred=tf.constant(value=True, dtype=tf.bool),
            true_fn=(lambda: self.summary_episode_reward(episode_reward=episode_reward)),
            false_fn=(lambda: self.assign_episode_reward(episode_reward=episode_reward))
        )
it works.

Another weird thing is I run the op in model.observe, but the error comes the first time do the q_model.update then I use dqn

Reply:
I think I see what you are trying to do but I am not sure it works the way you are trying to implement the change.  The summary ops are handled separately from the computational graph.  They will be fired I think after each episode is complete, I don't think there is a way to not pass the "episode_reward" to the summary when there is a non-terminal event.

Reply:
@Mazecreator I try to do the thing is to add up episode reward when non-terminal observation, and when terminal observation, summary it, so I try to return the assign_add op to session.run in model.observe, I'm a starter on tensorflow, so I'm not sure my code is right,  in the false_fn of tf.cond I only want to pass the assign_add op to model.observe, you can review the code in the PR, and thanks very much for your help

Reply:
The summary system is build into TF with a form of a callback.   Once you append it to the summaries() will be updated each episode regardless  of you tf.cond.  You  have formed an  operation in the TF graph and it will continue to be executed by the core TF RUN() callback.  The python code gets called once during the creation of the graph and builds the graph.  I don't think  this code is working the way you expect and I don't know of a way to fix it (so it only records reward upon terminal).  It just doesn't feel like it will fit into the structure right now with the summaries.

@AlexKuhnle  or @michaelschaarschmidt  might be able to chime in, but I don't see a way to only record the "episode_reward" upon terminal.  At best, you might be able to record 0 when not terminal and then episode_reward when terminal.  Not sure NA will work in TensorBoard (might throw and error).

Reply:
@Mazecreator  Thanks again, I'll try to figure out the TF summary mechanism and then modify my code
@AlexKuhnle @michaelschaarschmidt  I've notice that the default weight_init in Linear is Xavier of normal distribution
if self.weights_init is None:
    stddev = min(0.1, sqrt(2.0 / (x.shape[1].value + self.size)))
    self.weights_init = tf.random_normal_initializer(mean=0.0, stddev=stddev, dtype=tf.float32)
in the TF xavier_initializer, it doesn't have the min op, so could you explain the consideration about it, thanks!
https://www.tensorflow.org/api_docs/python/tf/contrib/layers/xavier_initializer
And also I think it's need MSRA initialization for PReLU according
https://arxiv.org/pdf/1502.01852.pdf

Reply:
@yellowbee686, the reason for this is that we at some point experimented with very small layer sizes in a toy example, which means that the `stddev` is rather big. We found that limiting it to `0.1` helped there. For "normal" layer sizes this restriction doesn't change the value at all, or at least not by much...

Reply:
@AlexKuhnle Thanks, that make sense, then I'll try to add the MSRA initialization and revert the episode reward code, which is apparently wrong with my misunderstand about TF summary.